### PR TITLE
Don't add volume on display from containers not worn

### LINKED ISSUE #41092: Maximum inventory volume is wrong
# Describe the bug

Exactly what it says on the tin - maximum available volume in inventory window shows wrong numbers. Like, way, way off the charts.

# Steps To Reproduce

- to be added later. Probably just put some empty non-rigid container(s?) inside others, that's my current guess.

# Expected behavior

Volume calculations returning correct result.

# Screenshots

cataclysm-tiles_2020_06_05_21_48_34_328
I guess it just summed all of my containers' max possible volumes? Or something?

# Versions and configuration

- OS: Windows 7x64
- Game Version: 0.E-2899-g0ff7d77 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Aftershock [aftershock],
    Magiclysm [magiclysm],
    Alternative Map Key [alt_map_key],
    sees-player icon, retrodays [sees_player_retro],
    Stats Through Skills [StatsThroughSkills]
]

### PR DESCRIPTION
**Pull request by meladath**
#### Summary

SUMMARY: Bugfixes "No longer add volume for every container inside every worn item unless that container is rigid"

#### Purpose of change

Fixes #41092.

#### Describe the solution

Character::volume_capacity and volume_carried adds the total_container_capacity of worn items, and weapon. However it also adds the total_container_capacity of all of the contents of those items too, even if they are rigid. Surely this is incorrect, as this will lead to capacities being added that cannot be utilised?

#### Describe alternatives you've considered

Spending a long time reading the rest of the code to figure out if this is some kind of mind trick or something.

#### Testing

Spawned in game, noticed my inventory volume had dropped by the correct amount considering nested containers do not increase your total carrying capacity.
I have a hiking backpack on, with a travelpack inside, inventory screens shows I have 65.50L max volume (currently only have scabbard, SWAT armor and hiking backpack containers worn) which is correct.

Tried spawning naked with a backpack, sheath, combat knife and glass jar. Glass jar inside backpack correctly increased current and total volume by .5L. backpack inside backpack correctly increases current volume by 2L and does not increase total volume. Sheathing the combat knife inside or outside the backpack does not gain any volume advantage, as the volume is added to the sheath's volume as it is non-rigid. More things that need to have JSON added to them. (sheaths, scabbard etc)

EDIT: After the most recent changes I have noticed a bug where books seem to reduce the amount of volume displayed you can carry. Please don't merge until I can take a look at this.

EDIT2: Rewritten PR into a recursive function that works out free space in order to calculate values.

### CODE REVIEW

**Comment by Resok** [comment]
Won't we get into a situation here where your 'displayed' volume is not relevent to actually picking up items through 'e'xamining a nearby tile to pick up items?

IE: user has their entire backpack filled with many empty bottles or gallon jugs, and their 'apparent' available volume grows with each, even though when they go to pick up an item they can't pickup any new items without manually inserting into each bottle or jug.

**Comment by meladath** [comment]
> 
> 
> Won't we get into a situation here where your 'displayed' volume is not relevent to actually picking up items through 'e'xamining a nearby tile to pick up items?
> 
> IE: user has their entire backpack filled with many empty bottles or gallon jugs, and their 'apparent' available volume grows with each, even though when they go to pick up an item they can't pickup any new items without manually inserting into each bottle or jug.

Good question. In short, no, since bottles/gallon jugs are rigid, the capacity will still be added, it will simply be added to both current and total volume.

for example, a backpack full of pastic bottles would show your volume as 15.00L/30.00L

**Comment by Resok** [comment]
> Ideally in the future there will be systems in place that will automate a lot of this inventory tetris but that is outside the scope of this PR.

Fair enough. I'm merely posing the question of if we should even be surfacing the rigid containers inside the worn/held inventory items or not since these are taking up inventory in your actual 'available' inventory as it's displayed. Surfacing these could easily be added back in when we have a practical system in place that auto-inserts into available sub-pockets but as it stands now either picking up with e or advanced inventory both won't expose any of the UI communicated 'available' volume.

Should we even be showing empty pockets within pockets as 'available' inventory since it creates a clear disconnect to general player behavior for picking things up?

**Comment by CodeBandit** [comment]
I don't think pockets inside pockets should even count as available volume.

**Comment by anothersimulacrum** [comment]
Why not? I can still put things in that box I have in my backpack, so that volume shouldn't all be shown as gone.
Non-rigid pockets, sure, but rigid ones should count.

**Comment by Resok** [comment]
> Why not? I can still put things in that box I have in my backpack, so that volume shouldn't all be shown as gone.
> Non-rigid pockets, sure, but rigid ones should count.

I think the problem is UI visualization. In the general UI that also shows up for examining to pickup - if the items won't automatically shift into containers within containers, then the values surfaced are incorrect on a practical level if we show rigid containers within worn containers.

If you're manually inserting into a container within another container, then that container shows it's available volume properly, but how does that relate to the 'at a glance' view that is your general inventory or examining a nearby tile to pick something up? (or with advanced inventory for that matter). It's currently setting a false pretense and is confusing in the UI.

It's also unclear on a surface level to the user as to why it would be showing the contents of one container type but not another.

Isn't the inventory or pickup UI essentially showing you the available space you can use to pickup items and place them into your currently worn/wielded containers?

**Comment by Resok** [comment]
Maybe containers within containers as available volume should only show up when using the new heirarchy view of the inventory? Since at least that presents a context that is clearer to the user.

Honestly it would be far more useful if the containers within your worn inventory showed some kind of held vs. max volume in the UI elements. Simply showing us that we supposedly have some amount of available inventory space, but not WHERE that space is available isn't terribly useful.

**Comment by CodeBandit** [comment]
Because having rigid containers inside backpacks doesn't give you more space. Putting a gallon jug in a backpack doesn't give you that backpack plus the gallon jug's volume, but that's how it would display. Imo, rigid containers themselves should count towards volume, anything inside them should not.

**Comment by meladath** [comment]
> 
> 
> Because having rigid containers inside backpacks doesn't give you more space. Putting a gallon jug in a backpack doesn't give you that backpack plus the gallon jug's volume, but that's how it would display. Imo, rigid containers themselves should count towards volume, anything inside them should not.

An interesting idea. Because then in my example if you have a backpack full of bottles you have 14.96/15.00L which instantly tells the player they cant pick anything else up, which is correct, however they can pick something up if the pickup can be contained inside the container.... again, I think its a UI problem and may require considerable reworking.

Switching it so it does this, won't be hard at all. What do people think?

**Comment by CodeBandit** [comment]
One possible solution could be an additional field displayed under capacity showing the available "rigid" space you have available in your worn containers. I have no particular qualms with the current implementation since it would be more intuitive for the player experience -though a little on the unrealistic side, depending on how the player chooses to interpret capacity.

**Comment by kevingranade** [comment]
We can resolve this if we flip around how it's calculated, what we can do is calculate "volume of worn items" minus "volume occupied by nested items", as that gives us "ideal amount of volume we can place in inventory".

**Comment by meladath** [comment]
PR now rewritten into a recursive function that appears to work with basic testing.

File: src/item_contents.h [hunk 1]
```diff
@@ -100,18 +103,29 @@ class item_contents
...
         units::volume total_container_capacity() const;
+
+        // Gets the total volume of every is_standard_type container
+        units::volume total_standard_capacity() const;
+
         units::volume remaining_container_capacity() const;
         units::volume total_contained_volume() const;
+
+        // gets all pockets contained in this item
+        ret_val<std::vector<item_pocket>> item_contents::get_all_contained_pockets() const;
```
Comment (kevingranade) [inline src/item_contents.h:114]:
```suggestion
        ret_val<std::vector<item_pocket>> get_all_contained_pockets() const;
```

File: src/item.h [hunk 2]
```diff
@@ -716,6 +716,9 @@ class item : public visitable<item>
          * ammo, magazines, weapons, etc.
          */
         units::volume get_total_capacity() const;
+
+        // recusive function that checks pockets for remaining free space
+        units::volume item::check_for_free_space( const item *it ) const;
```
Comment (kevingranade) [inline src/item.h:721]:
```suggestion
        units::volume check_for_free_space( const item *it ) const;
```

**Pull request merged**